netket/optimizer/sr/sr_onthefly.py

```diff
@@ -282,3 +300,199 @@ def fun(W, σ):
         res, _ = nkjax.tree_ravel(res)
 
     return res
+
+
+@struct.dataclass
+class JacobianSMatrix:
+    """
+    Semi-lazy representation of an S Matrix behaving like a linear operator.
+
+    The Jacobian O is computed on initialisation, but not S, which can be
+    computed by calling :code:`to_dense`.
+    The details on how the ⟨S⟩⁻¹⟨F⟩ system is solved are contaianed in
+    the field `sr`.
+    """
+
+    sr: SRLazy
+    """Parameters for the solution of the system."""
+
+    O: jnp.ndarray
+    """Gradients O_ij = ∂log ψ(σ_i)/∂p_j of the neural network 
+    for all samples σ_i at given values of the parameters p_j
+    Average <O_j> subtracted for each parameter
+    Divided through by sqrt(#samples) to normalise S matrix
+    If scale is not None, normalised to unit magnitude"""
+
+    scale: Optional[jnp.ndarray] = None
+    """If not None, gives scale factors with which O is normalised"""
+
+    x0: Optional[PyTree] = None
+    """Optional initial guess for the iterative solution."""
+    
+    def __matmul__(self, vec: Union[PyTree, jnp.ndarray]
+    ) -> Union[PyTree, jnp.ndarray]:
+        return jacsmatrix_mat_treevec(self, vec)
+
+    def __rtruediv__(self, y):
+        return self.solve(y)
+
+    def solve(self, y: PyTree, x0: Optional[PyTree] = None) -> PyTree:
+        """
+        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.
+
+        Args:
+            y: the vector y in the system above.
+            x0: optional initial guess for the solution.
+
+        Returns:
+            x: the PyTree solving the system.
+            info: optional additional informations provided by the solver. Might be
+                None if there are no additional informations provided.
+        """
+        if x0 is None:
+            x0 = self.x0
+        return jacsmatrix_solve(self,y,x0)
+
+    @jax.jit
+    def to_dense(self) -> jnp.ndarray:
+        """
+        Convert the lazy matrix representation to a dense matrix representation.
+
+        Returns:
+            A dense matrix representation of this S matrix.
+        """
+        return jnp.transpose(jnp.conj(self.O)) @ self.O \
+            + self.sr.diag_shift * jnp.eye(self.O.shape[1])
+
+def makeJacobianSMatrix(
+        apply_fun: Callable[[PyTree, jnp.ndarray], jnp.ndarray],
+        params: PyTree,
+        samples: jnp.ndarray,
+        sr: SRLazy,
+        model_state: Optional[PyTree] = None,
+        x0: Optional[PyTree] = None,
+        mode: str = "holomorphic",
+        rescale_shift: bool = False
+):
+    if rescale_shift:
+        O, scale = gradients(apply_fun, params, samples, model_state,
+                             mode, True)
+        np.savetxt('dum', 1/scale[:,jnp.newaxis])
```

Comment:
leftover?

Reply:
The print statement is a leftover indeed. I've remembered it just as I left from work... the branching is of course necessary to properly handle scale-invariant regularisation.